### PR TITLE
Run one JRuby worker on CI; allow to fail

### PR DESCRIPTION
**Pull request by bf4**
Ref:
- https://github.com/rails/rails/commit/0e8c04529159522aadbfe4fe71ea3d326df07d52
- https://github.com/rails/rails/issues/11700#issuecomment-21955029
## Update 2016-03-09

Attempts to download and install a binary https://github.com/rails/rails/pull/23499#discussion_r55489009

### CODE REVIEW

**Comment by rafaelfranca** [comment]
Is anyone going to work to get it fixed? Because if not I prefer to not add it back, unless this information is useful for someone.

**Comment by bf4** [comment]
There was a community-based effort earlier. That's why I made is allow failures.  Imo jruby is worth it

B mobile phone

**Comment by rafaelfranca** [comment]
I agree it is worth, but if no one is working on that we will keep this build in the allowed failure forever. If that is the case I prefer to not add it.

**Comment by bf4** [comment]
Ah, so you'd want to see a core-team champion and/or some trustworthy external group to be responsible.  

I can help find people, but I'm not sure if Rails would have a preferred group format/structure. 

Eg google group, org team, rails-docs-like repo to work off of, etc.

B mobile phone

**Comment by rafaelfranca** [comment]
@bf4 that would be awesome. We can invite these people to our community basecamp project. Could you send me an email so I can invite you to this room?

**Comment by bf4** [comment]
screen shot 2016-02-05 at 3 51 51 pm

If this goes in Rails in any way, it needs to run JRuby against a random sampling of the `GEM` directives.  Otherwise, it's still a waste of resources  --^ :eyes:

**Comment by gaurish** [comment]
@rafaelfranca I tried this with #11700, we made some progress.so I am interested in this

File: .travis.yml [hunk 1]
```diff
@@ -25,11 +25,19 @@ env:
...
+  # - 2.3.0
+  # - ruby-head
 matrix:
+  include:
+    - rvm: jruby-9.0.5.0
+      jdk: oraclejdk8
+      env:
+        - "JRUBY_OPTS='-J-Xmx1024M' -Xcompat.version=2.0 --server --Xcompile.invokedynamic=false -Xcli.debug=true --debug"
+        # Test a random part of Rails. Something is better than nothing, eh?
+        - 'GEM=$(ruby -e "print %w[railties ap ac am amo as av aj ar:postgresql ar:mysql2 ar:sqlite3 aj:integration].sample")'
```
Comment (bf4) [inline .travis.yml]: one solution to reduce overhead but ensure it's run regularly, randomly select one gem to run against

**Comment by bf4** [comment]
fwiw, if you want to retweet https://twitter.com/hazula/status/699431957327904768

**Comment by rafaelfranca** [comment]
Other way to work on this is to enable one component at time and get its test passing. Like I did in https://github.com/rails/rails/compare/enable-jruby

**Comment by bf4** [comment]
Or run JRuby on a different CI service? :smiling_imp:

**Comment by kares** [comment]
a random test will not be very useful. instead (if the core team does not want running the whole thing) run the parts that are close to :green_heart: and potentially get those cleaned up for JRuby. e.g. running the activesupport suite only. more useful than running some random part - say when a regression pops up due a change.

**Comment by gaurish** [comment]
Please see #23927

File: .travis.yml [hunk 2]
```diff
@@ -28,8 +28,15 @@ rvm:
   - 2.3.0
   - ruby-head
 matrix:
+  include:
+    - rvm: jruby-9.0.4.0 # Latest compiled version in http://rubies.travis-ci.org
+      jdk: oraclejdk8
+      env:
+        - "JRUBY_OPTS='-Xcompat.version=2.0 --server -Xcompile.invokedynamic=false'"
```
Comment (bf4) [inline .travis.yml]: per https://github.com/rails/rails/pull/16613 https://github.com/rails/rails/pull/17088 this should probably be `-J-Xmx1024M -Xcompat.version=2.0 --dev` but I'd want to verify

 cc @headius @robin850

Comment (kares) [inline .travis.yml]: `--dev -J-Xmx1024M'` is enough compat.version is deprecated on 9K

Comment (bf4) [inline .travis.yml]: Thanks @kares No wonder I couldn't find docs for it anymore or much in the code! https://twitter.com/hazula/status/705225976926572546

**Comment by robin850** [comment]
Hi there !

> I can help find people, but I'm not sure if Rails would have a preferred group format/structure.

A big :+1: here to put more efforts in making the test suite green on JRuby ; I didn't have much time to put in open source recently but this is definitely something that I want to work on when I'm able to.

If I can be on any help/advise, feel free to ping me on issues concerning JRuby or even Rubinius compatibility.

-----------------------
Here are the already known issues that make the suite to fail on JRuby:

* https://github.com/jruby/jruby/issues/3138
* https://github.com/jruby/jruby/issues/1958
* https://github.com/jruby/jruby/issues/1842
* https://github.com/jruby/jruby/issues/1840
* https://github.com/jruby/jruby/issues/1835
* https://github.com/jruby/jruby/issues/1806
* https://github.com/jruby/jruby/issues/1795
* https://github.com/jruby/jruby/issues/1769

And as for Rubinius (if anyone is interested):

* https://github.com/rubinius/rubinius/issues/3363 (IIRC, this one is pretty important ; the Active Record suite doesn't run at all because of this bug)
* https://github.com/rubinius/rubinius/issues/3334
* https://github.com/rubinius/rubinius/issues/2934 (Well, this one is pretty nebulous, this seems like there's a larger problem with autoloading and thread-safety on Rubinius)

(Sorry if some of these issues are out of date or don't concern the Rails code base anymore as it has evolved since they were reported).

------------------------

Also, during my Google Summer of Code, I had the idea to write [a guide to explain how to write code that works well on different rubies](http://robin.alwaysdata.net/rails/write_jruby_and_rubinius_compliant_code.html) ; I post it here as it may save you some time but it's a draft, it's not perfect and there *may* be inaccurate things so be careful.
There are also a lot of missing things ; I can try to update it when I have free time. Anyway, the **key** thing that's missing in this guide and that can seem obvious is : avoid mental compilations, as much as possible ! It's horrible the time you can loose when you're trying to guess where the problem's coming from and you can't figure it out because of a bad assumption.

Hope this can be of any help for you folks !

File: .travis.yml [hunk 3]
```diff
@@ -28,6 +28,12 @@ rvm:
   - 2.3.0
   - ruby-head
 matrix:
+  include:
+    - rvm: jruby-9.0.4.0 # Latest compiled version in http://rubies.travis-ci.org
```
Comment (rafaelfranca) [inline .travis.yml]: Hmmm. seems like this version was also compiled.

Comment (bf4) [inline .travis.yml]: I think it's just downloading it from S3, not compiling.  JRuby also provides its own binaries. Why @travis-ci doesn't use them is a mystery to me. Got any friends @travis-ci that can help us figure this out?  https://github.com/bf4/rails/commit/6a314d44c439b0a1ff702f1ec0871dfc0bc705da it appears 9.0.5.0 is now available https://github.com/travis-ci/travis-ci/issues/5733

I'm doing the incredibly crazy

```yaml
before_install:
  - '[ "$JRUBY_OPTS" != "" ] && export JRUBY_OPTS="--dev -J-Xmx1024M" && rvm mount -r http://rubies.travis-ci.org/ubuntu/12.04/x86_64/jruby-9.0.5.0.tar.gz --verify-downloads 1 --binary --ignore-gemsets && rvm use jruby-9.0.5.0 --binary --ignore-gemsets && gem install bundler --conservative'
matrix:
  include:
      jdk: oraclejdk8
      rvm: default
      env:
          - "JRUBY_OPTS='--dev -J-Xmx1024M'"
          - "GEM='ap'"
```

**Comment by bf4** [comment]
@robin850 if you have any insight into getting jruby to install faster, much appreciated.  I'm mostly out of ideas... https://github.com/rails/rails/pull/23499#discussion_r55489009 https://github.com/bf4/rails/commit/f04e78b3a2caa48d9298c6cc912457a849b5b988

**Comment by bf4** [comment]
@rafaelfranca Build is green. JRuby took 9 minutes compared to about 2 on other rubies, but is still faster than railties which takes 20+ min cc @robin850 @kaspth @gaurish

**Comment by rafaelfranca** [comment]
LGTM. Could you squash the commits?

**Comment by bf4** [comment]
@rafaelfranca Updated

**Comment by bf4** [comment]
@jeremy 

> This build is routinely hanging and failing

I've seen this before.  I'll try to figure it out.

**Comment by bf4** [comment]
Oh, this was on my list of JRuby failures I made a while back in #17587 https://gist.github.com/bf4/bfb08415fde03f46601f

**Pull request merged**